File: server/src/routes/agent.js
// server/src/routes/agent.js
import express                from 'express';
import { HumanMessage, AIMessage } from '@langchain/core/messages';
import { createCustomerAgent }     from '../agents/customer-agent.js';

const router = express.Router();

const agentApp = createCustomerAgent();

router.post('/stream', async (req, res) => {
  const { message, history = [] } = req.body;

  if (!message) return res.status(400).json({ error: 'message 不能为空' });

  res.setHeader('Content-Type', 'text/event-stream; charset=utf-8');
  res.setHeader('Cache-Control', 'no-cache');
  res.setHeader('Connection', 'keep-alive');
  res.setHeader('X-Accel-Buffering', 'no');

  const send = (type, data) =>
    res.write(`data: ${JSON.stringify({ type, ...data })}\n\n`);

  try {
    // 将历史记录转为消息对象（排除最后一条，避免重复）
    const historyMessages = history.slice(0, -1).map((m) =>
      m.role === 'user' ? new HumanMessage(m.content) : new AIMessage(m.content)
    );

    const result = await agentApp.invoke({
      messages: [...historyMessages, new HumanMessage(message)],
    });

    // 从消息列表提取工具调用步骤
    const msgs = result.messages;
    for (let i = 0; i < msgs.length; i++) {
      const msg = msgs[i];
      if (msg.tool_calls?.length) {
        for (const tc of msg.tool_calls) {
          const toolResult = msgs[i + 1];
          send('step', {
            tool:        tc.name,
            toolInput:   tc.args,
            observation: toolResult?.content ?? '',
          });
        }
      }
    }

    const finalMsg = msgs[msgs.length - 1];
    send('answer', { content: finalMsg.content });
    send('done',   {});
    res.end();
  } catch (err) {
    console.error('[Agent Error]', err.message);
    send('error', { content: '处理请求时出错，请重试' });
    res.end();
  }
});

export default router;

File: server-py/app/routers/agent.py
import json

from fastapi import APIRouter
from fastapi.responses import JSONResponse, StreamingResponse
from langchain_core.messages import AIMessage, HumanMessage
from pydantic import BaseModel

from app.agents.customer_agent import create_customer_agent

router = APIRouter()

_agent_app = create_customer_agent()


class ChatMessage(BaseModel):
    role: str
    content: str


class AgentRequest(BaseModel):
    message: str
    history: list[ChatMessage] = []


def _send(event_type, data):
    return f"data: {json.dumps({'type': event_type, **data}, ensure_ascii=False)}\n\n"


@router.post("/stream")
async def agent_stream(req: AgentRequest):
    if not req.message:
        return JSONResponse(status_code=400, content={"error": "message 不能为空"})

    async def event_generator():
        try:
            # 将历史记录转为消息对象（排除最后一条，避免重复）
            history_messages = [
                HumanMessage(content=m.content) if m.role == "user" else AIMessage(content=m.content)
                for m in req.history[:-1]
            ]

            result = await _agent_app.ainvoke(
                {"messages": history_messages + [HumanMessage(content=req.message)]}
            )

            # 从消息列表提取工具调用步骤
            msgs = result["messages"]
            for i, msg in enumerate(msgs):
                tool_calls = getattr(msg, "tool_calls", None)
                if tool_calls:
                    for tc in tool_calls:
                        tool_result = msgs[i + 1] if i + 1 < len(msgs) else None
                        yield _send(
                            "step",
                            {
                                "tool": tc["name"],
                                "toolInput": tc["args"],
                                "observation": tool_result.content if tool_result else "",
                            },
                        )

            final_msg = msgs[-1]
            yield _send("answer", {"content": final_msg.content})
            yield _send("done", {})
        except Exception as err:
            print(f"[Agent Error] {err}")
            yield _send("error", {"content": "处理请求时出错，请重试"})

    return StreamingResponse(event_generator(), media_type="text/event-stream")
